datasync/chaindatafetcher/chaindata_fetcher.go

```diff
@@ -120,88 +124,102 @@ func (f *ChainDataFetcher) Start(server p2p.Server) error {
+	if f.fetchingStarted {
+		return errors.New("fetching is already started")
 	}
 
 	f.chainSub = f.blockchain.SubscribeChainEvent(f.chainCh)
 	currentBlock := f.blockchain.CurrentHeader().Number.Uint64()
-	if err := f.startRangeFetching(uint64(f.checkpoint), currentBlock, requestTypeAll); err != nil {
-		return err
-	}
 
-	f.started = true
+	f.fetchingStopCh = make(chan struct{})
+	f.fetchingWg.Add(1)
+	go func() {
+		defer f.fetchingWg.Done()
+		f.makeRequests(uint64(f.checkpoint), currentBlock, requestTypeAll, f.fetchingStopCh)
+	}()
+	f.fetchingStarted = true
 	return nil
 }
 
 func (f *ChainDataFetcher) stopFetching() error {
-	if !f.started {
-		return errors.New("the chaindata fetcher is not running")
+	if !f.fetchingStarted {
+		return errors.New("fetching is not running")
 	}
 
 	f.chainSub.Unsubscribe()
-	f.started = false
+	close(f.fetchingStopCh)
+	f.fetchingWg.Wait()
+	f.fetchingStarted = false
 	return nil
 }
 
 func (f *ChainDataFetcher) startRangeFetching(start, end uint64, reqType requestType) error {
-	if f.rangeStarted {
-		return errors.New("the chaindata fetcher is already started with range")
-	}
-	f.rangeStarted = true
-	defer func() { f.rangeStarted = false }()
-
-	// TODO-ChainDataFetcher parallelize the following codes
-	for i := start; i < end; i++ {
-		e, err := f.makeChainEvent(i)
-		if err != nil {
-			return err
-		}
-
-		f.reqCh <- newRequest(reqType, e)
-		// TODO-ChainDataFetcher add stop logic while processing the events.
+	if f.rangeFetchingStarted {
+		return errors.New("range fetching is already started")
 	}
+	f.rangeFetchingStopCh = make(chan struct{})
+	f.rangeFetchingWg.Add(1)
+	go func() {
+		defer f.rangeFetchingWg.Done()
+		f.makeRequests(start, end, reqType, f.rangeFetchingStopCh)
+	}()
+	f.rangeFetchingStarted = true
 	return nil
 }
 
 func (f *ChainDataFetcher) stopRangeFetching() error {
-	// TODO-ChainDataFetcher add logic for stopping
+	if !f.rangeFetchingStarted {
+		return errors.New("range fetching is not running")
+	}
+	close(f.rangeFetchingStopCh)
+	f.rangeFetchingWg.Wait()
+	f.rangeFetchingStarted = false
 	return nil
 }
 
-// TODO-ChainDataFetcher push down this logic to handleRequest
 func (f *ChainDataFetcher) makeChainEvent(blockNumber uint64) (blockchain.ChainEvent, error) {
 	var logs []*types.Log
 	block := f.blockchain.GetBlockByNumber(blockNumber)
 	receipts := f.blockchain.GetReceiptsByBlockHash(block.Hash())
 	for _, r := range receipts {
 		logs = append(logs, r.Logs...)
 	}
-	fct := "fastCallTracer"
-	results, err := f.debugAPI.TraceBlockByNumber(context.Background(), rpc.BlockNumber(block.Number().Int64()), &cn.TraceConfig{
-		Tracer: &fct,
-	})
-	if err != nil {
-		return blockchain.ChainEvent{}, err
-	}
 	var internalTraces []*vm.InternalTxTrace
-	for _, r := range results {
-		// TODO-ChainDataFetcher Assume that the input parameters are valid always.
-		internalTraces = append(internalTraces, r.Result.(*vm.InternalTxTrace))
+	if blockNumber != 0 {
```

Comment:
Without checking this, the trace API returns an error due to the `not found 0x00...00` error

Reply:
Why don't you return an error when blockNumber is 0

Reply:
I made it do traceBlock when only any transaction exists